crazy_rl/multi_agent/jax/surround/surround.py

```diff
@@ -0,0 +1,203 @@
+                if other_agent != agent:
+                    obs_agent = jnp.append(obs_agent, state.agent_location[other_agent])
+
+            obs = obs.at[agent].set(obs_agent)
+
+        return jdc.replace(state, observations=obs)
+
+    @override
+    @partial(jit, static_argnums=(0,))
+    def _compute_action(self, actions, state):
+        # Actions are clipped to stay in the map and scaled to do max 20cm in one step
+        return jdc.replace(
+            state,
+            agent_location=jnp.clip(state.agent_location + actions * 0.2, jnp.array([-self.size, -self.size, 0]), self.size),
+        )
+
+    @override
+    @partial(jit, static_argnums=(0,))
+    def _compute_reward(self, state):
+        # Reward is the mean distance to the other agents plus a maximum value minus the distance to the target
+
+        return jdc.replace(
+            state,
+            rewards=jnp.any(state.truncations)
+            * (
+                # mean distance to the other agents
+                jnp.array(
+                    [
+                        jnp.sum(self.norm(state.agent_location[agent] - state.agent_location))
+                        for agent in range(self.num_drones)
+                    ]
+                )
+                * 0.05
+                / (self.num_drones - 1)
+                # a maximum value minus the distance to the target
+                + 0.95 * (2 * self.size - self.norm(state.agent_location - self._target_location))
+            )
+            # negative reward if the drones crash
+            + jnp.any(state.terminations) * -10 * jnp.ones(self.num_drones),
+        )
+
+    @override
+    @partial(jit, static_argnums=(0,))
+    def _compute_terminated(self, state):
+        # collision with the ground and the target
+        terminated = jnp.logical_or(
+            state.agent_location[:, 2] < 0.2, jnp.linalg.norm(state.agent_location - self._target_location) < 0.2
+        )
+
+        for agent in range(self.num_drones):
+            distances = self.norm(state.agent_location[agent] - state.agent_location)
+
+            # collision between two drones
+            terminated = terminated.at[agent].set(
+                jnp.logical_or(terminated[agent], jnp.any(jnp.logical_and(distances > 0.001, distances < 0.2)))
+            )
+
+        return jdc.replace(
+            state, terminations=((jnp.any(state.truncations) - 1) * jnp.any(terminated)) * jnp.ones(self.num_drones)
+        )
+
+    @override
+    @partial(jit, static_argnums=(0,))
+    def _compute_truncation(self, state):
+        return jdc.replace(state, truncations=(state.timestep == 100) * jnp.ones(self.num_drones))
+
+
+if __name__ == "__main__":
+    parallel_env = Surround(
+        num_drones=5,
+        render_mode=None,
+        init_flying_pos=jnp.array([[0, 0, 1], [2, 1, 1], [0, 1, 1], [2, 2, 1], [1, 0, 1]]),
+        target_location=jnp.array([[1, 1, 2.5]]),
+    )
+
+    parallel_env.state = parallel_env.reset(parallel_env.state)
```

Comment:
Yep, so as I said before with having state as class attribute. I would much rather have it returned to the outside:
```python
env = Surround(num_drones=3, render_mode=None, init_flying_pos=..., target_location=...)
state = env.reset() # maybe provide a seed or RNGKey from jax

done = False
while not done:
    state = env.step(state, actions)
    done = any(state.terminateds) or any(state.truncateds)
```